{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/amyvandiver/Box/Nanopore/Timp_data\n"
     ]
    }
   ],
   "source": [
    "cd Timp_data/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!gunzip 210617_HV_mito_targseq_2.gup5.fq.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!gunzip 190226_HV_mitoenrich.gup5.fq.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "x= 'HV/210617_HV_mito_targseq_2.gup5'\n",
    "fq = x + '.fq'\n",
    "bam= x + '_rot.bam'\n",
    "\n",
    "x1= 'HV/190226_HV_mitoenrich.gup5'\n",
    "fq1 = x1 + '.fq'\n",
    "bam1= x1 + '_rot.bam'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6273 reads\n",
      "769 reads\n"
     ]
    }
   ],
   "source": [
    "!echo $(( $(wc -l < $fq) / 4 )) reads\n",
    "!echo $(( $(wc -l < $fq1) / 4 )) reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref=\"chrM_rot.fa\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def align(fastq, ref,outname):\n",
    "    #align\n",
    "    !minimap2 -ax map-ont {ref} -t 11 {fastq} --MD| samtools view -uF 4 | samtools sort -o {outname}\n",
    "    !samtools index {outname}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[M::mm_idx_gen::0.003*1.35] collected minimizers\n",
      "[M::mm_idx_gen::0.005*3.96] sorted minimizers\n",
      "[M::main::0.005*3.93] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.006*3.84] mid_occ = 2\n",
      "[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.006*3.76] distinct minimizers: 3103 (100.00% are singletons); average occurrences: 1.000; average spacing: 5.340\n",
      "[M::worker_pipeline::2.069*5.25] mapped 6273 sequences\n",
      "[M::main] Version: 2.17-r941\n",
      "[M::main] CMD: minimap2 -ax map-ont -t 11 --MD chrM_rot.fa HV/210617_HV_mito_targseq_2.gup5.fq\n",
      "[M::main] Real time: 2.072 sec; CPU: 10.857 sec; Peak RSS: 0.310 GB\n",
      "[M::mm_idx_gen::0.006*0.79] collected minimizers\n",
      "[M::mm_idx_gen::0.007*3.00] sorted minimizers\n",
      "[M::main::0.007*2.99] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.008*2.95] mid_occ = 2\n",
      "[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.008*2.91] distinct minimizers: 3103 (100.00% are singletons); average occurrences: 1.000; average spacing: 5.340\n",
      "[M::worker_pipeline::0.347*6.37] mapped 769 sequences\n",
      "[M::main] Version: 2.17-r941\n",
      "[M::main] CMD: minimap2 -ax map-ont -t 11 --MD chrM_rot.fa HV/190226_HV_mitoenrich.gup5.fq\n",
      "[M::main] Real time: 0.349 sec; CPU: 2.213 sec; Peak RSS: 0.071 GB\n"
     ]
    }
   ],
   "source": [
    "align(fq, ref,bam)\n",
    "align(fq1, ref,bam1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1761\n",
      "1761\n",
      "457\n",
      "457\n"
     ]
    }
   ],
   "source": [
    "!samtools view -F 260 -F 2048 {bam} chrM_rot -c\n",
    "!samtools view -F 260 -F 2048 {bam} -c\n",
    "\n",
    "!samtools view -F 260 -F 2048 {bam1} chrM_rot -c\n",
    "!samtools view -F 260 -F 2048 {bam1} -c\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "!samtools merge HV/HV_all.bam {bam1} {bam} -f\n",
    "!samtools index HV/HV_all.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "bamall=\"HV/HV_all.bam\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2218\r\n"
     ]
    }
   ],
   "source": [
    "!samtools view -F 260 -F 2048 {bamall} chrM_rot -c"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HV/210617_HV_mito_targseq_2.gup5_rot.bam\n",
      "HV/190226_HV_mitoenrich.gup5_rot.bam\n"
     ]
    }
   ],
   "source": [
    "print(bam)\n",
    "print(bam1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number mito reads: 1761\n",
      "Mito reads greater than 10kb: 1339\n",
      "Mean mito read length: 13622.72345258376\n",
      "Number mito reads: 457\n",
      "Mito reads greater than 10kb: 294\n",
      "Mean mito read length: 11673.487964989059\n"
     ]
    }
   ],
   "source": [
    "!samtools view -F 260 -F 2048 HV/210617_HV_mito_targseq_2.gup5_rot.bam chrM_rot| awk '{print length($10)}' >lengths.txt\n",
    "lengths=pd.read_csv('lengths.txt',sep='\\t',names=[\"length\"])\n",
    "print(\"Number mito reads: \" + str(len(lengths)))\n",
    "print(\"Mito reads greater than 10kb: \" + str(lengths.length[lengths.length >10000].count()))\n",
    "print(\"Mean mito read length: \" + str(lengths.length.mean()))\n",
    "\n",
    "!samtools view -F 260 -F 2048 HV/190226_HV_mitoenrich.gup5_rot.bam chrM_rot| awk '{print length($10)}' >lengths.txt\n",
    "lengths=pd.read_csv('lengths.txt',sep='\\t',names=[\"length\"])\n",
    "print(\"Number mito reads: \" + str(len(lengths)))\n",
    "print(\"Mito reads greater than 10kb: \" + str(lengths.length[lengths.length >10000].count()))\n",
    "print(\"Mean mito read length: \" + str(lengths.length.mean()))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number mito reads: 2218\n",
      "Mito reads greater than 10kb: 1633\n",
      "Mean mito read length: 13221.100090171325\n"
     ]
    }
   ],
   "source": [
    "!samtools view -F 260 -F 2048 HV/HV_all.bam chrM_rot| awk '{print length($10)}' >lengths.txt\n",
    "lengths=pd.read_csv('lengths.txt',sep='\\t',names=[\"length\"])\n",
    "print(\"Number mito reads: \" + str(len(lengths)))\n",
    "print(\"Mito reads greater than 10kb: \" + str(lengths.length[lengths.length >10000].count()))\n",
    "print(\"Mean mito read length: \" + str(lengths.length.mean()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "outdir=\"bcftools/\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def np_bcf(bam, ref, outdir, label, af=0.05):\n",
    "    chrM_reads=!samtools idxstats {bam} | grep 'chrM_rot' | cut -f3\n",
    "    cov_thresh=int(chrM_reads[0])/10\n",
    "    \n",
    "    bcfpile=outdir+label+\".bcfpile.vcf\"\n",
    "    \n",
    "    !bcftools mpileup -a \"AD,ADF,ADR\" -Q 0 -B -d 100000 --threads 8 -Ou -f {ref} -r chrM_rot {bam} | bcftools filter -Ou -e 'INFO/DP[*] < {cov_thresh}' | \\\n",
    "        bcftools filter -i \"(FORMAT/AD[0:1]) > {af}*(INFO/DP[0])\" -o {bcfpile}\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[mpileup] 1 samples in 1 input files\n",
      "[mpileup] 1 samples in 1 input files\n"
     ]
    }
   ],
   "source": [
    "np_bcf(bamall,ref,outdir,\"HV_all\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[bgzip] No such file or directory: bcftools/HV_all.bcfpile.vcf\n",
      "[tabix] the index file exists. Please use '-f' to overwrite.\n"
     ]
    }
   ],
   "source": [
    "!bgzip bcftools/HV_all.bcfpile.vcf\n",
    "!tabix -p vcf bcftools/HV_all.bcfpile.vcf.gz\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tHV/HV_all.bam\n",
      "chrM_rot\t95\t.\tG\tA,C,T\t0\tPASS\tDP=1930;I16=1563,210,135,22,28834,583890,2421,46387,106203,6.372e+06,9420,565200,44317,1.10779e+06,3925,98125;QS=0.922134,0.0753306,0.00215047,0.000385159;VDB=0;SGB=-0.693147;RPB=0.694488;MQB=0.999381;MQSB=0.999045;BQB=0.43895;MQ0F=0\tPL:ADF:ADR:AD\t0,255,236,255,255,255,255,255,255,255:1563,127,8,0:210,19,0,3:1773,146,8,3\n",
      "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tHV/HV_all.bam\n",
      "chrM_rot\t11966\t.\tG\tA,C,T\t0\tPASS\tDP=1506;I16=1135,261,100,10,26874,614042,2708,82178,83760,5.0256e+06,6600,396000,34900,872500,2747,68609;QS=0.909395,0.0900633,0.000372162,0.000169165;VDB=0;SGB=-0.693147;RPB=0.931367;MQB=1;MQSB=1;BQB=2.19547e-06;MQ0F=0\tPL:ADF:ADR:AD\t0,255,243,255,255,255,255,255,255,255:1135,99,0,1:261,9,1,0:1396,108,1,1\n"
     ]
    }
   ],
   "source": [
    "!bcftools view -r chrM_rot:95 bcftools/HV_all.bcfpile.vcf.gz | grep -v '##'\n",
    "!bcftools view -r chrM_rot:11966 bcftools/HV_all.bcfpile.vcf.gz | grep -v '##'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pysam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "myloc=[94,11965]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "region=[min(myloc), max(myloc)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "coverage at base 94 = 1981\n",
      "coverage at base 11965 = 1546\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"4\" halign=\"left\">read</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>unique</th>\n",
       "      <th>top</th>\n",
       "      <th>freq</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>geno</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>-A</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>ae61f350-df33-4d14-85f2-8a271f0dd7c5</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-G</th>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>2e15f762-2f78-4bb5-80e3-09c50ed6f9ca</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A-</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>14ed5fa5-1efd-460c-89a7-92a743d8d34d</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AA</th>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>0ef4438f-b128-4d56-8bc6-3de63af8eebb</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AG</th>\n",
       "      <td>87</td>\n",
       "      <td>87</td>\n",
       "      <td>80468a7f-e1e2-45bb-9040-c2804aec9329</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CG</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>38f4bfe8-1e87-4860-8de7-1ce2dbbf30c0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G-</th>\n",
       "      <td>32</td>\n",
       "      <td>32</td>\n",
       "      <td>7655e956-ffa8-4e8d-bcb5-63946b1a0285</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GA</th>\n",
       "      <td>80</td>\n",
       "      <td>80</td>\n",
       "      <td>ece892cc-dca2-4770-adf2-fb25a0a5c784</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GC</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>a77fdb52-3d26-448a-98fe-45c894d72eb4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GG</th>\n",
       "      <td>1183</td>\n",
       "      <td>1183</td>\n",
       "      <td>efd58fb7-9836-4d83-add2-14c247215673</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GT</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2325e8e6-1069-47d9-bece-f20818a5a853</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TG</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>241f933b-2073-4cb9-bcba-786aa661d1aa</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      read                                                  \n",
       "     count unique                                   top freq\n",
       "geno                                                        \n",
       "-A       3      3  ae61f350-df33-4d14-85f2-8a271f0dd7c5    1\n",
       "-G      31     31  2e15f762-2f78-4bb5-80e3-09c50ed6f9ca    1\n",
       "A-       2      2  14ed5fa5-1efd-460c-89a7-92a743d8d34d    1\n",
       "AA      19     19  0ef4438f-b128-4d56-8bc6-3de63af8eebb    1\n",
       "AG      87     87  80468a7f-e1e2-45bb-9040-c2804aec9329    1\n",
       "CG       5      5  38f4bfe8-1e87-4860-8de7-1ce2dbbf30c0    1\n",
       "G-      32     32  7655e956-ffa8-4e8d-bcb5-63946b1a0285    1\n",
       "GA      80     80  ece892cc-dca2-4770-adf2-fb25a0a5c784    1\n",
       "GC       1      1  a77fdb52-3d26-448a-98fe-45c894d72eb4    1\n",
       "GG    1183   1183  efd58fb7-9836-4d83-add2-14c247215673    1\n",
       "GT       1      1  2325e8e6-1069-47d9-bece-f20818a5a853    1\n",
       "TG       3      3  241f933b-2073-4cb9-bcba-786aa661d1aa    1"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bammy=pysam.AlignmentFile(bamall, \"rb\")\n",
    "\n",
    "mapped=bammy.pileup(\"chrM_rot\", region[0], region[1],min_base_quality=0, max_depth=1000000)\n",
    "readvariants=dict()\n",
    "for pileupcolumn in mapped:\n",
    "    if pileupcolumn.pos in myloc:\n",
    "        print(\"coverage at base %s = %s\" %(pileupcolumn.pos, pileupcolumn.n))\n",
    "        for pileupread in pileupcolumn.pileups:\n",
    "            readname=pileupread.alignment.query_name\n",
    "            if pileupread.is_del:\n",
    "                base=\"-\" \n",
    "            else:\n",
    "                base=pileupread.alignment.query_sequence[pileupread.query_position]\n",
    "            if readname in readvariants:\n",
    "                readvariants[readname]=readvariants[readname]+base\n",
    "            else:\n",
    "                readvariants[readname]=base\n",
    "\n",
    "readvariants={key: value for (key, value) in readvariants.items() if len(value) == len(myloc)}\n",
    "\n",
    "\n",
    "geno=pd.DataFrame.from_dict(readvariants, orient='index',columns=[\"geno\"])\n",
    "geno.index.name = \"read\"\n",
    "geno.reset_index(inplace=True)\n",
    "geno.groupby(\"geno\").describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "geno.to_csv(\"HV/geno.csv\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "snp1 = []\n",
    "for x in geno.geno:\n",
    "    snp1.append(str(x)[0])\n",
    "\n",
    "snp2 = []\n",
    "for x in geno.geno:\n",
    "    snp2.append(str(x)[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "first=geno.read[geno.snp1==\"A\"]\n",
    "second=geno.read[geno.snp2==\"A\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "102"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(first)\n",
    "len(second)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "geno['read'] = pd.Series(snp1)\n",
    "geno['snp2']=pd.Series(snp2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame(\n",
    "[snp1,snp2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1447, 4)"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "geno.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "import upsetplot\n",
    "from upsetplot import from_contents"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>snp1</th>\n",
       "      <th>snp2</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">True</th>\n",
       "      <th>False</th>\n",
       "      <td>f520816f-252d-4dab-bc0d-057323ec6111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>True</th>\n",
       "      <td>99a3b461-e068-469f-9e34-ec5006ccb920</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>False</th>\n",
       "      <td>709d1470-8dec-4c36-837e-2f5ff643b1d6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>False</th>\n",
       "      <td>1575d16b-3580-4b3a-a56c-3dc444dc7b03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>False</th>\n",
       "      <td>eaf2f32f-a165-4703-b563-e4700c4ef42a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">False</th>\n",
       "      <th>True</th>\n",
       "      <td>6e08848c-31c4-4595-a5a2-70614d37e9a3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>True</th>\n",
       "      <td>ae956b3a-f1a9-496d-88b2-ace4bc5972de</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>True</th>\n",
       "      <td>e0aaca77-1c7a-47a3-9050-9920af0e8f63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>True</th>\n",
       "      <td>b9a67e7c-9f45-422e-8580-4ddfdf327dcc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>True</th>\n",
       "      <td>71cc8054-e26e-4b61-8caf-e82e710cb964</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>191 rows Ã— 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               id\n",
       "snp1  snp2                                       \n",
       "True  False  f520816f-252d-4dab-bc0d-057323ec6111\n",
       "      True   99a3b461-e068-469f-9e34-ec5006ccb920\n",
       "      False  709d1470-8dec-4c36-837e-2f5ff643b1d6\n",
       "      False  1575d16b-3580-4b3a-a56c-3dc444dc7b03\n",
       "      False  eaf2f32f-a165-4703-b563-e4700c4ef42a\n",
       "...                                           ...\n",
       "False True   6e08848c-31c4-4595-a5a2-70614d37e9a3\n",
       "      True   ae956b3a-f1a9-496d-88b2-ace4bc5972de\n",
       "      True   e0aaca77-1c7a-47a3-9050-9920af0e8f63\n",
       "      True   b9a67e7c-9f45-422e-8580-4ddfdf327dcc\n",
       "      True   71cc8054-e26e-4b61-8caf-e82e710cb964\n",
       "\n",
       "[191 rows x 1 columns]"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from upsetplot import from_contents\n",
    "test = from_contents({'snp1': first, 'snp2': second})\n",
    "test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "f=plt.figure(figsize=(25,10))\n",
    "plt.title('Clustering Dendogram')\n",
    "dendrogram(link,truncate_mode='lastp', labels=idx)\n",
    "plt.show\n",
    "\n",
    "f.savefig(\"HV/clustering_trunc.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAMIAAADgCAYAAABRhbkMAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+WH4yJAAATHElEQVR4nO3debQcZZnH8e+PsCQBQsgw5CJBiBhxIAhCBASXKDjGYT2gAuPMIDKoM7KNg8ri0Yh6FCfDmHFDFBERQQhxLtthmWBwYU0AgYhIhigGQkIgEIRASPLMH2+16Vxu7q3qru7qvvl9zunT6eqqt58k/XS9VfW+TykiMNvQbVR1AGadwIlghhPBDHAimAFOBDPAiWAGOBGGlClTpgTgx/of6+VEGEKWLl1adQhdy4lghhPBDHAimAFOBDPAibBB6OnpQVLuR09PT9Uht50TYQOwePHilq4/FDgRzHAimAFOBDPAiWAGOBHMACeCGeBEMAOcCFaCoXDBzolgTRsKF+ycCB1A0r9JmifpQUmXSRouaYykmyU9kj1vXXWcQ5kToWKStgdOASZFxERgGHAMcAYwKyImALOy19YiToTOsDEwQtLGwEjgCeBw4OLs/YuBIyqKbYOwcdUBbOgi4nFJ04DHgBXATRFxk6SxEbEoW2eRpG37217SR4GPAowdO5bZs2eXEldZ7VTVfn8mT5683vfk2qfVyvr+VwFHA88CVwIzgG9GxOi69ZZFxIDHCZMmTYo5c+b09xmF4yryvWh1+yVab6DuGlXvIGBBRDwVEa8AM4H9gcWStgPInpdUGOOQ50So3mPAfpJGKv20Hgg8BFwNHJetcxzQW1F8GwQfI1QsIu6UNAO4B1gF3AtcAGwBXCHpBFKyfKC6KIc+HyMMIT5GGJSPEcwG4kQww4lgBjgRzAAnghngRCiVpB0lHZT9eYSkLauOyfJxIpRE0omkoRHfzRaNA/6nuoisCCdCeT4BHAAsB4iIR4B+B8pZ53EilOfliFhZe5ENqfbVyi7hRCjPrZLOIs0reA9pFOk1FcdkOTkRynMG8BTwAPAx4PqIOLvakCwvD7orz4eAyyPie7UFkg6JiGsrjMly8h6hPN8Afinpb+qWnVNVMFaME6E8C4CPADMk1YZMFx+WaZVw16g8ERH3SHoncJmkfUkVKawLeI9QntpE+6XAe0mnTidWGpHl5kQoSUQcXPfnNRHxqYjwv2+XcNeoSZK+HhGnSbqGfi6gRcRhFYRlBTkRmndJ9jyt0iisKU6EJkXE3Oz51tqyrFbRDhFxf2WBWSHuw5ZE0mxJoySNAX4DXCTpvKrjsnycCOXZKiKWA0cCF0XE3qTiXYOSNFrSDEm/k/SQpLe6GnZ7ORHKs3FWke6DQNFhFdOBGyLijcAepAJfrobdRk6E8pwD3AjMj4i7Jb0OeGSwjSSNAt4BXAgQESsj4llcDbutfLBckoi4kjT0uvb6UeCoHJu+jjRq9SJJewBzgVMBV8MumathdzBJk4A7gAOy8o/TSbPcTnY17NK50l0HWwgsjIg7s9czgL1wNey2ciJULCKeBP4kaZds0YHAb3E17LbyMUJJJG1GOibYibp/14jIMyfhZOBSSZsCjwLHk36kXA27TZwI5ekFniMd7L5cZMOIuA+Y1M9bB5YQl+XgRCjPuIiYUnUQ1hgfI5TnNkm7Vx2ENcZ7hPK8DfiwpAWkrpFIs9beVG1YlocToTzvqzoAa5y7RiWJiD8Co4FDs8fobJl1ASdCSSSdClxKqne6LfBjSSdXG5Xl5a5ReU4A9o2IFwAknQvcTqp3ZB3Oe4TyCFhd93o1rmvUNbxHKM9FwJ2Sfpa9PoJsaLV1PidCSSLiPEmzSadRBRwfEfdWG5Xl5URokqRREbE8m6v8h+xRe29MRDxTVWyWnxOheT8BDiGNMaofZK/s9euqCMqKcSI0KSIOyZ7HVx2LNc5njUoiaVaeZdaZvEdokqThwEhgm6zkSu2U6SjgNZUFZoU4EZr3MeA00pd+LmsTYTnwraqCsmKcCE2KiOnAdEknR4SvIncpHyOUZ42k+qoTW0v61yoDsvycCOU5MSvMBUBELANOrDAeK8CJUJ6NVFfgR9IwYNMK47ECfIxQnhtJVSfOJ11I+zhwQ7UhWV5OhPJ8hnQG6V9IZ45uAr6fd+NsDzIHeDwiDsmGbPyUVB7mD8AHs+6WtYC7RiWJiDXAD4GzI+KoiPhuRKweZLN6p5KqYNe4GnYbORFKIukw4D6y7pCkPSVdnXPbccDBrLsHcTXsNnIilOfzwD7As/CXol075dz268CngTV1y9aphk2a/mkt4mOE8qyKiOeKVoaWdAiwJCLmSppc9ENdFj4/l4VvA0kXsrYvfxRwCrBJRHx8kO2+AvwjsAoYThqjNBN4CzA5uzfCdsDsiNhl/S25LHwOLgvfBicDu5GKe11GGmt02mAbRcSZETEuInYCjgFuiYh/wNWw28pdo5JExIvA2cDZ2anQzSPipSaa/Cquht023iOURNJPstvLbg7MAx6W9KkibUTE7LqJPk9HxIERMSF79pTPFnIilGfX7PayRwDXA68l9f2tCzgRyrOJpE1IidAbEa+w7hxm62BOhPKcTxoKsTnwC0k7kg6YrQv4YLkEkjYCFkfE9nXLHgPeVV1UVoT3CCXIxhmd1GdZRMSqikKygpwI5blZ0umSdpA0pvaoOijLx12j8nwke/5E3TIX+OoSToSSuMBXd3PXqCSSRkr6rKQLstcTsgF11gWcCOW5CFgJ7J+9Xgh8qbpwrAgnQnl2joivAa8ARMQKfKOQruFEKM9KSSPIriZL2pk0EtW6gA+WyzOVNE1zB0mXAgcAx1cakeXmRChJRNwkaS6wH6lLdGpELK04LMvJXaOSSJqVDZ2+LiKujYilLgvfPbxHaJLLwg8NToTmuSz8EOBEaJLLwg8NToSSRMQ3JO1PqmW0cd3yH1UWlOXmRCiJpEuAnUnV7mqlHgNwInQBJ0J5JpHmLXt6Zhfy6dPyPAj0VB2ENcZ7hPJsA/xW0l3UDa2IiMMG2kjSDqTuUw+p9ukFETHdZeHby4lQnqkNbrcK+PeIuEfSlsBcSTcDHyaVhf+qpDNIpSQ/U0qk9ipOhJJExK0NbrcIqFW9fl7SQ8D2pLLwk7PVLgZm40RoGSdCkyQ9T//1i0Sawz+qQFs7AW8G7qRPWXhJ/ZaFdzXs/FwNuwtI2gK4FfhyRMyU9GxE1N+udllEbD1QG0O1GnZPTw+LFy/Ovf7YsWN58skn+3vL1bA7WVYh7yrg0oiYmS1enJWDJ3teUlV8VSuSBI2sD06EymW3pL0QeCgizqt7y2Xh28jHCNU7gFQs+AFJ92XLzsJl4dvKiVCxiPgV6++7HtjOWDZk7hqZ4UQwA5wIZoATwQxwIpgBTgQzwIlgBjgRzAAnghngRDADnAhmgBPBDHAimAFOBDPAiWAGOBHMACeCGeBEMAOcCGaAE8EMcCKYAU6EjiZpiqSHJc3PCgFbizgROpSkYaSbEb4P2BU4VtKu1UY1dDkROtc+wPyIeDQiVgKXkypkWwu4wFfn2h74U93rhcC+fVeqr4YNzAMm9l2n1YWeu719cCJ0sv6q373qGxERFwAXtD6coc1do861ENih7vU44ImKYhnynAid625ggqTxkjYFjiFVyLYWcNeoQ0XEKkknATcCw4AfRMS8isMasnzHHDPcNTIDnAhmgBPBDHAimAFOBDPAiWAGOBHMgJISobe394Yy2jGrSll7hG1KasfaQNJwSaOzxyYltjtM0qGSbpG0SNLjkn4m6W3ZjdXLav/nWfsLJc2UdECz7Tc9xEJSTJs2DUldcYk6Ipr+D+lWkkYC2wGb91m+HFgUES830fY44BagB9iy7q3DgIOAuZIOjYjnm2j/58DYPu0fDrwHmJO1/+dG2vcxwgZC0hbAzvRJgswo4PWSNmuw7dHAbcB41v2SQvqObUGaS3GjpMI/vpK2Bm4Hdhqg/f2AG7KZfYU5EbqApD0l3S5pnqT7JR1dcHsBr6X/OQ41w1h32HcRJwN/zcA9jOHA7sChDbR/Cqn7PVj7ewCHNNC+E6FLvAj8U0TsBkwBvp79Cue1Ffm6wSMljSgSWPYLfArpiziYLYBCRQiyPUiR9j9TpP0aJ0KLSdpc0nWSfiPpQUlHS/qDpC9IukfSA5LemK07VdIl2cHmI5JOBIiI30fEI9mfnwCWkH6B8+rbnShrXYAdgSLJs3fBA9udgCJdtrcUWPcvnAitNwV4IiL2iIiJQO1U89KI2Av4DnB63fpvAg4G3gp8TtJr6huTtA+wKfB/BWIo8sUrejJhM2BNgfU3KvgZRdsfJqnw99qJ0HoPAAdJOlfS2yPiuWz5zOx5LulXr6Y3IlZExFLSWZJ9am9I2g64BDg+Iop8OYqcDVpZYF2ARaTEzGtZwdifKNj+0wXbB5wILRcRvwf2JiXEVyR9Lnur9uVczbr9976noQNA0ijgOuCzEXFHwTCW5VxvNfDcoGvViYhngVn0U1igHy8B5xdsfxnpB6El7dc4EVos69q8GBE/BqYBew2yyeHZBa+/AiYDd2dzln8G/CgiriwaQ1YX6Zkcqy5p5NcU+DKwIsd6q4FvN9D+lwq0/50G2ncitMHuwF2S7gPOJv2nDuQu0i//HcAXs4PjDwLvAD4s6b7ssWfBOB5n4D3Dkoh4qmCbAETEbcCnSWe3+l0le+/9EfF4A+3/GjgzR/tHZv9exUVEUw8gpk2bFlkwHf9o9u/bygcwFTi9xZ8xglQaZgLweuA1wGYltf1eYE72pXw2e6wAbgL2LqH9KaRjqhfq/k9XkAoc7NVM200PsYgI9fb2zomISc22Za0XEStINZNa0faNpKvHbwDeQDrbc39ElPJ5EXED6erxLsDvssUTymjf5Vw6SERMrTqGMkQ6QfD7Frb/cO1SRFlJ5mMEM5wIZoATwQxwIpgBTgQzoMGzRpLWGaF41VVXbdR3mVk7FPnexQCz47xHMMPXETZEWwJHkW5QuIZ0Jfhqio86HUxtqHXZc9m3BN5f9/oo4BqajL/RRFhe/2LBggWvWtbBRlUdQEU2As4h3W9tNWvnLv8ZmA58kQZHbvb5jK2BMawdOv0SacBf3hGw6zOMFOM/s+78hG8A/w18gSZuoeU9wobjh8Df8uopj1tkz1NJ84IHGxS4PhuTJu/3nTswnDSeaSvgjzS2hxBpHsa7WX/855CS8NwG2vcxQheZSbrL5hUNbHsUqeTJyAHWGUmahP/mBtqHVBxgoAk0m5MSohEfAN7F4PF/kjTatzAnQveYztrbyBb1Sfov49LXpqRkKGoE+eYtb0Xq4hRVJP6TGmjfidAGI4ErSXV/7gSOBB4EzgJ+SZp3MCFb90xSP/da4F7guLp2bgUaKY61LWkkaB7DaKzcSt6KGiIlQxHbkeox5TEMOKJg+4AToR0OAp4E9icVufrfbPnTwNuB7wOn1q0/kXRW5EBS6ZOeJj9/NMXOqGxG8V/tIseaRdsuGv8IihcgcCK0wW9JUy7PIVWmqJ1dq90q9j5S/7rmOtaeafkFab5zM56j2OT3laSzSkWsKrBu0bafA4rUZ32JBg7InQitN580zXIe6RRfrQBVocn7TVgMPJpz3TXA9Q18Rt4J/0Hx0+xPkE4S5LGa1K0szInQej2kqYs/JR3wDjbX+GBS92QMqet0Twkx/BdpeuNgXiKdly/qxWzbwSyn2N6j5jzWP1+53svANxto34nQBrsBs4FfA58CvjbI+nOBGaQSKeeSji8gzcu9hNTN+h3pGCKvnwK/YuAv04vARcDdBdqt9xgD9+VXkAoINOIyUhHggSpZvAh8jwZ/OHxBrfVmZY96E+v+fC/wd3Wv55Nqffb13iZiCOBYUjmZD5G6QLXTnS+QfhD/E/iPJj7jFVIXbAzpwlatX7+SdGJgGY1389aQKnmcBxyTtVO7sPYC6eD4a9n7DWk0EdYZpjB+/PhbgXc2GoS1xSrgNNJB+9+TLjytJp3SnUG+rtNgVgNPZY9hdcvK8ArpGsdUUvwTSX+nWvx5uk7r5T1CZ/lKGz7jGRrsRxdUVgL09TSNHccMyMcIZjS4R+g7waG3t3fNQJMezDqd9whmgLJSesU2km5g3TtpbgMsLSuoFhse6T4FZn/RUCK8qhGpa0o+dlOs1j7uGpnhRDADykuEhueKVqCbYrU2KeUYwazbuWtkRo5EkPQDSUskPVi3bIykm7N7Ad8saeu6986UNF/Sw5KaGShWKklTspjmSyp002sb+vLsEX5IumVPvTOAWRExgTSy8gwASbuSRgfW7hD/7ezO7JXKYvgW8D5SYatjs1jNgByJEBG/4NV3ZDwcuDj788WsnTB9OHB5RLwcEQtIQ4r3oXr7APMj4tFId5i8nBSrGdD4McLYiFgEkD1vmy3fnnWn1S3MllWtU+OyDlH2wXJ/1QM64bRUp8ZlHaLRRFgsaTuA7HlJtnwhsEPdeuNIk6+r1qlxWYdoNBGuZm3xqeOA3rrlx0jaTNJ4UuGqu5oLsRR3AxMkjc/uYn8Ma8upmA0+H0HSZaQJ49tIWgh8HvgqcIWkE0iTtj8AEBHzJF1BquWzCvhERLRqplJuEbFK0kmkCfDDgB9ExLyKw7IO4ivLZvjKshngRDADnAhmgBPBDHAimAFOBDPAiWAGOBHMAPh/K3nMgf1St2EAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 224x256 with 4 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "from upsetplot import UpSet\n",
    "UpSet(test, subset_size='count').plot()\n",
    "plt.show\n",
    "plt.savefig(\"upset.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "coverage at base 94 = 110380\n",
      "coverage at base 11965 = 105811\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"4\" halign=\"left\">read</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>unique</th>\n",
       "      <th>top</th>\n",
       "      <th>freq</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>geno</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>--</th>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>c28d4740-3598-4648-8c15-95b95d34f788</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-A</th>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>254a6509-b88e-4680-83b8-35cb295c2d8c</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-C</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>8ee9330d-6a62-447a-8830-ee61c87959c1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-G</th>\n",
       "      <td>1128</td>\n",
       "      <td>1128</td>\n",
       "      <td>477d6da7-0527-47cd-937a-a1f9fc90c5d6</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-T</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>c6f63ac5-8791-4ed0-8ee4-8429b477ca2c</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A-</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>f2edbe60-c1eb-4f3f-852d-27739af4f8a0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AA</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1e4e192b-c066-4289-9776-f43fb45312f2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AG</th>\n",
       "      <td>242</td>\n",
       "      <td>242</td>\n",
       "      <td>83ec4303-6d12-4d69-b050-c5d0fd395d3f</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C-</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>c8e9a1ab-2cf6-4c31-bf7c-edfaf5fc5807</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CA</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>e4735b5f-1e19-4385-87b0-9170e9f93164</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CG</th>\n",
       "      <td>218</td>\n",
       "      <td>218</td>\n",
       "      <td>a1d7d525-4ea2-4edb-97f0-45cd35cdd61b</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G-</th>\n",
       "      <td>961</td>\n",
       "      <td>961</td>\n",
       "      <td>c7530746-9e5d-4728-becd-c7f976c5e899</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GA</th>\n",
       "      <td>670</td>\n",
       "      <td>670</td>\n",
       "      <td>e14298e6-2c26-4493-82fd-f64f84a10f28</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GC</th>\n",
       "      <td>72</td>\n",
       "      <td>72</td>\n",
       "      <td>e0700c53-e1b4-410f-9373-6922d82d187a</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GG</th>\n",
       "      <td>95721</td>\n",
       "      <td>95721</td>\n",
       "      <td>39e9cd96-8551-4a3e-a433-bfb54ca2bdaa</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GT</th>\n",
       "      <td>77</td>\n",
       "      <td>77</td>\n",
       "      <td>651c605f-71c4-4429-8206-ae8bda001afa</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T-</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>6a5e8589-5d31-41fd-8ef4-8174d1708034</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TA</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>b7aae0bc-227d-43bd-ac30-4f2fadb652bc</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TG</th>\n",
       "      <td>75</td>\n",
       "      <td>75</td>\n",
       "      <td>0116e948-81d4-4d5e-a681-7cfb898a11da</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       read                                                  \n",
       "      count unique                                   top freq\n",
       "geno                                                         \n",
       "--       19     19  c28d4740-3598-4648-8c15-95b95d34f788    1\n",
       "-A       10     10  254a6509-b88e-4680-83b8-35cb295c2d8c    1\n",
       "-C        1      1  8ee9330d-6a62-447a-8830-ee61c87959c1    1\n",
       "-G     1128   1128  477d6da7-0527-47cd-937a-a1f9fc90c5d6    1\n",
       "-T        2      2  c6f63ac5-8791-4ed0-8ee4-8429b477ca2c    1\n",
       "A-        1      1  f2edbe60-c1eb-4f3f-852d-27739af4f8a0    1\n",
       "AA        1      1  1e4e192b-c066-4289-9776-f43fb45312f2    1\n",
       "AG      242    242  83ec4303-6d12-4d69-b050-c5d0fd395d3f    1\n",
       "C-        4      4  c8e9a1ab-2cf6-4c31-bf7c-edfaf5fc5807    1\n",
       "CA        4      4  e4735b5f-1e19-4385-87b0-9170e9f93164    1\n",
       "CG      218    218  a1d7d525-4ea2-4edb-97f0-45cd35cdd61b    1\n",
       "G-      961    961  c7530746-9e5d-4728-becd-c7f976c5e899    1\n",
       "GA      670    670  e14298e6-2c26-4493-82fd-f64f84a10f28    1\n",
       "GC       72     72  e0700c53-e1b4-410f-9373-6922d82d187a    1\n",
       "GG    95721  95721  39e9cd96-8551-4a3e-a433-bfb54ca2bdaa    1\n",
       "GT       77     77  651c605f-71c4-4429-8206-ae8bda001afa    1\n",
       "T-        1      1  6a5e8589-5d31-41fd-8ef4-8174d1708034    1\n",
       "TA        1      1  b7aae0bc-227d-43bd-ac30-4f2fadb652bc    1\n",
       "TG       75     75  0116e948-81d4-4d5e-a681-7cfb898a11da    1"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "myloc=[94,11965]\n",
    "bammy=pysam.AlignmentFile(\"GM12878/190115_cas9enrich_mito1_multif5.gup5.rotM.bam\", \"rb\")\n",
    "\n",
    "mapped=bammy.pileup(\"chrM_rot\", myloc[0], myloc[1],min_base_quality=0, max_depth=1000000)\n",
    "readvariants=dict()\n",
    "for pileupcolumn in mapped:\n",
    "    if pileupcolumn.pos in myloc:\n",
    "        print(\"coverage at base %s = %s\" %(pileupcolumn.pos, pileupcolumn.n))\n",
    "        for pileupread in pileupcolumn.pileups:\n",
    "            readname=pileupread.alignment.query_name\n",
    "            if pileupread.is_del:\n",
    "                base=\"-\" \n",
    "            else:\n",
    "                base=pileupread.alignment.query_sequence[pileupread.query_position]\n",
    "            if readname in readvariants:\n",
    "                readvariants[readname]=readvariants[readname]+base\n",
    "            else:\n",
    "                readvariants[readname]=base\n",
    "\n",
    "readvariants={key: value for (key, value) in readvariants.items() if len(value) == len(myloc)}\n",
    "\n",
    "\n",
    "geno=pd.DataFrame.from_dict(readvariants, orient='index',columns=[\"geno\"])\n",
    "geno.index.name = \"read\"\n",
    "geno.reset_index(inplace=True)\n",
    "geno.groupby(\"geno\").describe()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "coverage at base 94 = 31971\n",
      "coverage at base 11965 = 6958\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"4\" halign=\"left\">read</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>unique</th>\n",
       "      <th>top</th>\n",
       "      <th>freq</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>geno</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>--</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2b94e9cf-cac4-49a2-b7fd-eaeb9d474a71</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-G</th>\n",
       "      <td>44</td>\n",
       "      <td>44</td>\n",
       "      <td>7e1b9779-dd2b-42c7-ad29-aa6714255f9b</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AG</th>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>39162c90-bc9b-4c0b-8b02-90fb104e0576</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CG</th>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>c99f4b23-9f07-4239-b525-cedbe2fab56e</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G-</th>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>3c7ca8b9-f1e4-4128-afb6-3735117d90f5</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GA</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>00f1bbfb-6727-4c49-9f7d-408c26e2511e</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GG</th>\n",
       "      <td>2069</td>\n",
       "      <td>2069</td>\n",
       "      <td>b57ce83e-7f29-4073-95f8-82e5df3621a1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GT</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>f4517a50-b18c-4cd6-a2d8-d7e6ed4a0732</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      read                                                  \n",
       "     count unique                                   top freq\n",
       "geno                                                        \n",
       "--       1      1  2b94e9cf-cac4-49a2-b7fd-eaeb9d474a71    1\n",
       "-G      44     44  7e1b9779-dd2b-42c7-ad29-aa6714255f9b    1\n",
       "AG       9      9  39162c90-bc9b-4c0b-8b02-90fb104e0576    1\n",
       "CG       9      9  c99f4b23-9f07-4239-b525-cedbe2fab56e    1\n",
       "G-      15     15  3c7ca8b9-f1e4-4128-afb6-3735117d90f5    1\n",
       "GA       4      4  00f1bbfb-6727-4c49-9f7d-408c26e2511e    1\n",
       "GG    2069   2069  b57ce83e-7f29-4073-95f8-82e5df3621a1    1\n",
       "GT       1      1  f4517a50-b18c-4cd6-a2d8-d7e6ed4a0732    1"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "myloc=[94,11965]\n",
    "bammy=pysam.AlignmentFile(\"Muscle/bam/190226_UCLA_HD1.rot.bam\", \"rb\")\n",
    "\n",
    "mapped=bammy.pileup(\"chrM_rot\", myloc[0], myloc[1],min_base_quality=0, max_depth=1000000)\n",
    "readvariants=dict()\n",
    "for pileupcolumn in mapped:\n",
    "    if pileupcolumn.pos in myloc:\n",
    "        print(\"coverage at base %s = %s\" %(pileupcolumn.pos, pileupcolumn.n))\n",
    "        for pileupread in pileupcolumn.pileups:\n",
    "            readname=pileupread.alignment.query_name\n",
    "            if pileupread.is_del:\n",
    "                base=\"-\" \n",
    "            else:\n",
    "                base=pileupread.alignment.query_sequence[pileupread.query_position]\n",
    "            if readname in readvariants:\n",
    "                readvariants[readname]=readvariants[readname]+base\n",
    "            else:\n",
    "                readvariants[readname]=base\n",
    "\n",
    "readvariants={key: value for (key, value) in readvariants.items() if len(value) == len(myloc)}\n",
    "\n",
    "\n",
    "geno=pd.DataFrame.from_dict(readvariants, orient='index',columns=[\"geno\"])\n",
    "geno.index.name = \"read\"\n",
    "geno.reset_index(inplace=True)\n",
    "geno.groupby(\"geno\").describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "coverage at base 94 = 63356\n",
      "coverage at base 11965 = 59518\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"4\" halign=\"left\">read</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>unique</th>\n",
       "      <th>top</th>\n",
       "      <th>freq</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>geno</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>--</th>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "      <td>2b841124-8db0-4b32-9565-6b5c26f477f1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-A</th>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>7790b966-30b9-4ee9-b065-4973cc7c7d90</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-C</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>81319c8f-4f13-459e-a542-8a1f04b55f0f</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-G</th>\n",
       "      <td>1067</td>\n",
       "      <td>1067</td>\n",
       "      <td>edd9ea0b-d2fb-4ef4-9279-19b3cb7a59fe</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-T</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>c19fb68d-875d-407c-889d-dd705ee2c599</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A-</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>a78b0e59-139c-469a-9cbb-6b56dc9f35bf</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AA</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>d8d392ff-5a10-4718-936a-552ee52c6016</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AG</th>\n",
       "      <td>213</td>\n",
       "      <td>213</td>\n",
       "      <td>f5698bc2-d14f-4e9a-b33e-5ac861d2f1ea</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C-</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>30e3105d-feff-43e4-a241-686ebf5434e7</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CA</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>c180ec5b-071e-436b-b9c0-4b53f81b5aa4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CG</th>\n",
       "      <td>156</td>\n",
       "      <td>156</td>\n",
       "      <td>39d2d251-7421-4a75-bdcd-61b92bd7dcf0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G-</th>\n",
       "      <td>839</td>\n",
       "      <td>839</td>\n",
       "      <td>a77c034a-53db-4666-a819-33125a405d94</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GA</th>\n",
       "      <td>410</td>\n",
       "      <td>410</td>\n",
       "      <td>b38a2102-c1a5-4d43-a734-bd680df8ab82</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GC</th>\n",
       "      <td>97</td>\n",
       "      <td>97</td>\n",
       "      <td>647783dd-e4de-463b-88d5-ca7f2251ebaa</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GG</th>\n",
       "      <td>49404</td>\n",
       "      <td>49404</td>\n",
       "      <td>c34198c4-114e-4841-ac1d-15ffce814d52</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GT</th>\n",
       "      <td>88</td>\n",
       "      <td>88</td>\n",
       "      <td>c39a1e22-f254-4491-95de-840b7fbf5842</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T-</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1fce6e27-8bd7-4de0-9c83-3a8db99f63fb</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TG</th>\n",
       "      <td>59</td>\n",
       "      <td>59</td>\n",
       "      <td>a5a22a47-5d1e-40ce-a93d-df924cc65178</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TT</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>dc74056f-2e14-41ef-b129-7fcf5dc80305</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       read                                                  \n",
       "      count unique                                   top freq\n",
       "geno                                                         \n",
       "--       21     21  2b841124-8db0-4b32-9565-6b5c26f477f1    1\n",
       "-A       14     14  7790b966-30b9-4ee9-b065-4973cc7c7d90    1\n",
       "-C        3      3  81319c8f-4f13-459e-a542-8a1f04b55f0f    1\n",
       "-G     1067   1067  edd9ea0b-d2fb-4ef4-9279-19b3cb7a59fe    1\n",
       "-T        1      1  c19fb68d-875d-407c-889d-dd705ee2c599    1\n",
       "A-        5      5  a78b0e59-139c-469a-9cbb-6b56dc9f35bf    1\n",
       "AA        2      2  d8d392ff-5a10-4718-936a-552ee52c6016    1\n",
       "AG      213    213  f5698bc2-d14f-4e9a-b33e-5ac861d2f1ea    1\n",
       "C-        2      2  30e3105d-feff-43e4-a241-686ebf5434e7    1\n",
       "CA        1      1  c180ec5b-071e-436b-b9c0-4b53f81b5aa4    1\n",
       "CG      156    156  39d2d251-7421-4a75-bdcd-61b92bd7dcf0    1\n",
       "G-      839    839  a77c034a-53db-4666-a819-33125a405d94    1\n",
       "GA      410    410  b38a2102-c1a5-4d43-a734-bd680df8ab82    1\n",
       "GC       97     97  647783dd-e4de-463b-88d5-ca7f2251ebaa    1\n",
       "GG    49404  49404  c34198c4-114e-4841-ac1d-15ffce814d52    1\n",
       "GT       88     88  c39a1e22-f254-4491-95de-840b7fbf5842    1\n",
       "T-        2      2  1fce6e27-8bd7-4de0-9c83-3a8db99f63fb    1\n",
       "TG       59     59  a5a22a47-5d1e-40ce-a93d-df924cc65178    1\n",
       "TT        1      1  dc74056f-2e14-41ef-b129-7fcf5dc80305    1"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "myloc=[94,11965]\n",
    "bammy=pysam.AlignmentFile(\"Muscle/bam/190314_UCLA_old.rot.bam\", \"rb\")\n",
    "\n",
    "mapped=bammy.pileup(\"chrM_rot\", myloc[0], myloc[1],min_base_quality=0, max_depth=1000000)\n",
    "readvariants=dict()\n",
    "for pileupcolumn in mapped:\n",
    "    if pileupcolumn.pos in myloc:\n",
    "        print(\"coverage at base %s = %s\" %(pileupcolumn.pos, pileupcolumn.n))\n",
    "        for pileupread in pileupcolumn.pileups:\n",
    "            readname=pileupread.alignment.query_name\n",
    "            if pileupread.is_del:\n",
    "                base=\"-\" \n",
    "            else:\n",
    "                base=pileupread.alignment.query_sequence[pileupread.query_position]\n",
    "            if readname in readvariants:\n",
    "                readvariants[readname]=readvariants[readname]+base\n",
    "            else:\n",
    "                readvariants[readname]=base\n",
    "\n",
    "readvariants={key: value for (key, value) in readvariants.items() if len(value) == len(myloc)}\n",
    "\n",
    "\n",
    "geno=pd.DataFrame.from_dict(readvariants, orient='index',columns=[\"geno\"])\n",
    "geno.index.name = \"read\"\n",
    "geno.reset_index(inplace=True)\n",
    "geno.groupby(\"geno\").describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "coverage at base 94 = 108684\n",
      "coverage at base 11965 = 95305\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"4\" halign=\"left\">read</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>unique</th>\n",
       "      <th>top</th>\n",
       "      <th>freq</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>geno</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>--</th>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>2ee2993c-1d70-4f51-8742-aa5a05d2ece0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-A</th>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>b7793349-5fb6-4928-a4da-66dff8490554</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-C</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>75822757-8cf1-4b58-9309-859dd29e732e</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-G</th>\n",
       "      <td>1246</td>\n",
       "      <td>1246</td>\n",
       "      <td>37252a93-0b5e-4afc-a2f1-9b8c94677920</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-T</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>a371539f-4db9-47e5-8167-b92eee07a52a</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A-</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>b21d62dd-fb6a-49de-9269-8cd47e5812e5</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AA</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0cb46d10-cda8-48cc-9692-1f362f3bf805</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>d1561e25-6b77-49ba-aefc-c35c8e37a4b2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AG</th>\n",
       "      <td>258</td>\n",
       "      <td>258</td>\n",
       "      <td>3beba5ff-e3b2-4f39-8677-11c4df61286f</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C-</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>9881f62c-48c8-4a4a-9077-da998bfa285f</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CA</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>138f3737-5fda-4b56-8350-70dbc7043888</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CG</th>\n",
       "      <td>216</td>\n",
       "      <td>216</td>\n",
       "      <td>3f19c25c-4a69-489d-88ef-a0cb04afc557</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G-</th>\n",
       "      <td>911</td>\n",
       "      <td>911</td>\n",
       "      <td>7b649e13-b7e4-4000-9f78-dd86db465ddf</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GA</th>\n",
       "      <td>522</td>\n",
       "      <td>522</td>\n",
       "      <td>c2989e3e-0054-433f-8dab-41377f00cacf</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GC</th>\n",
       "      <td>102</td>\n",
       "      <td>102</td>\n",
       "      <td>22d216e6-0e56-42b8-9590-7c7c23a81e52</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GG</th>\n",
       "      <td>75812</td>\n",
       "      <td>75812</td>\n",
       "      <td>eaa26c97-60c1-4f84-a686-fe8ad76476d6</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GT</th>\n",
       "      <td>81</td>\n",
       "      <td>81</td>\n",
       "      <td>05223156-3110-4c9e-bcee-f663694bc863</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T-</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>49270610-2f46-4895-a22a-cc9bc51c38c7</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TA</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>08da06fe-6f62-4239-b6cd-62d2d1dece00</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TG</th>\n",
       "      <td>72</td>\n",
       "      <td>72</td>\n",
       "      <td>a2d695a2-5cb6-4810-8619-0c3fc4f652ed</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TT</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>c7cc8f74-5c39-4121-8314-7a1e08dca56e</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       read                                                  \n",
       "      count unique                                   top freq\n",
       "geno                                                         \n",
       "--       14     14  2ee2993c-1d70-4f51-8742-aa5a05d2ece0    1\n",
       "-A        7      7  b7793349-5fb6-4928-a4da-66dff8490554    1\n",
       "-C        1      1  75822757-8cf1-4b58-9309-859dd29e732e    1\n",
       "-G     1246   1246  37252a93-0b5e-4afc-a2f1-9b8c94677920    1\n",
       "-T        1      1  a371539f-4db9-47e5-8167-b92eee07a52a    1\n",
       "A-        4      4  b21d62dd-fb6a-49de-9269-8cd47e5812e5    1\n",
       "AA        2      2  0cb46d10-cda8-48cc-9692-1f362f3bf805    1\n",
       "AC        1      1  d1561e25-6b77-49ba-aefc-c35c8e37a4b2    1\n",
       "AG      258    258  3beba5ff-e3b2-4f39-8677-11c4df61286f    1\n",
       "C-        2      2  9881f62c-48c8-4a4a-9077-da998bfa285f    1\n",
       "CA        1      1  138f3737-5fda-4b56-8350-70dbc7043888    1\n",
       "CG      216    216  3f19c25c-4a69-489d-88ef-a0cb04afc557    1\n",
       "G-      911    911  7b649e13-b7e4-4000-9f78-dd86db465ddf    1\n",
       "GA      522    522  c2989e3e-0054-433f-8dab-41377f00cacf    1\n",
       "GC      102    102  22d216e6-0e56-42b8-9590-7c7c23a81e52    1\n",
       "GG    75812  75812  eaa26c97-60c1-4f84-a686-fe8ad76476d6    1\n",
       "GT       81     81  05223156-3110-4c9e-bcee-f663694bc863    1\n",
       "T-        1      1  49270610-2f46-4895-a22a-cc9bc51c38c7    1\n",
       "TA        1      1  08da06fe-6f62-4239-b6cd-62d2d1dece00    1\n",
       "TG       72     72  a2d695a2-5cb6-4810-8619-0c3fc4f652ed    1\n",
       "TT        1      1  c7cc8f74-5c39-4121-8314-7a1e08dca56e    1"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "myloc=[94,11965]\n",
    "bammy=pysam.AlignmentFile(\"Muscle/bam/190314_UCLA_young.rot.bam\", \"rb\")\n",
    "\n",
    "mapped=bammy.pileup(\"chrM_rot\", myloc[0], myloc[1],min_base_quality=0, max_depth=1000000)\n",
    "readvariants=dict()\n",
    "for pileupcolumn in mapped:\n",
    "    if pileupcolumn.pos in myloc:\n",
    "        print(\"coverage at base %s = %s\" %(pileupcolumn.pos, pileupcolumn.n))\n",
    "        for pileupread in pileupcolumn.pileups:\n",
    "            readname=pileupread.alignment.query_name\n",
    "            if pileupread.is_del:\n",
    "                base=\"-\" \n",
    "            else:\n",
    "                base=pileupread.alignment.query_sequence[pileupread.query_position]\n",
    "            if readname in readvariants:\n",
    "                readvariants[readname]=readvariants[readname]+base\n",
    "            else:\n",
    "                readvariants[readname]=base\n",
    "\n",
    "readvariants={key: value for (key, value) in readvariants.items() if len(value) == len(myloc)}\n",
    "\n",
    "\n",
    "geno=pd.DataFrame.from_dict(readvariants, orient='index',columns=[\"geno\"])\n",
    "geno.index.name = \"read\"\n",
    "geno.reset_index(inplace=True)\n",
    "geno.groupby(\"geno\").describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "coverage at base 94 = 108684\n",
      "coverage at base 11965 = 95305\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"4\" halign=\"left\">read</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>unique</th>\n",
       "      <th>top</th>\n",
       "      <th>freq</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>geno</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>--</th>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>2ee2993c-1d70-4f51-8742-aa5a05d2ece0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-A</th>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>b7793349-5fb6-4928-a4da-66dff8490554</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-C</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>75822757-8cf1-4b58-9309-859dd29e732e</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-G</th>\n",
       "      <td>1246</td>\n",
       "      <td>1246</td>\n",
       "      <td>37252a93-0b5e-4afc-a2f1-9b8c94677920</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-T</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>a371539f-4db9-47e5-8167-b92eee07a52a</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A-</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>b21d62dd-fb6a-49de-9269-8cd47e5812e5</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AA</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0cb46d10-cda8-48cc-9692-1f362f3bf805</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>d1561e25-6b77-49ba-aefc-c35c8e37a4b2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AG</th>\n",
       "      <td>258</td>\n",
       "      <td>258</td>\n",
       "      <td>3beba5ff-e3b2-4f39-8677-11c4df61286f</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C-</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>9881f62c-48c8-4a4a-9077-da998bfa285f</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CA</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>138f3737-5fda-4b56-8350-70dbc7043888</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CG</th>\n",
       "      <td>216</td>\n",
       "      <td>216</td>\n",
       "      <td>3f19c25c-4a69-489d-88ef-a0cb04afc557</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G-</th>\n",
       "      <td>911</td>\n",
       "      <td>911</td>\n",
       "      <td>7b649e13-b7e4-4000-9f78-dd86db465ddf</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GA</th>\n",
       "      <td>522</td>\n",
       "      <td>522</td>\n",
       "      <td>c2989e3e-0054-433f-8dab-41377f00cacf</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GC</th>\n",
       "      <td>102</td>\n",
       "      <td>102</td>\n",
       "      <td>22d216e6-0e56-42b8-9590-7c7c23a81e52</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GG</th>\n",
       "      <td>75812</td>\n",
       "      <td>75812</td>\n",
       "      <td>eaa26c97-60c1-4f84-a686-fe8ad76476d6</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GT</th>\n",
       "      <td>81</td>\n",
       "      <td>81</td>\n",
       "      <td>05223156-3110-4c9e-bcee-f663694bc863</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T-</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>49270610-2f46-4895-a22a-cc9bc51c38c7</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TA</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>08da06fe-6f62-4239-b6cd-62d2d1dece00</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TG</th>\n",
       "      <td>72</td>\n",
       "      <td>72</td>\n",
       "      <td>a2d695a2-5cb6-4810-8619-0c3fc4f652ed</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TT</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>c7cc8f74-5c39-4121-8314-7a1e08dca56e</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       read                                                  \n",
       "      count unique                                   top freq\n",
       "geno                                                         \n",
       "--       14     14  2ee2993c-1d70-4f51-8742-aa5a05d2ece0    1\n",
       "-A        7      7  b7793349-5fb6-4928-a4da-66dff8490554    1\n",
       "-C        1      1  75822757-8cf1-4b58-9309-859dd29e732e    1\n",
       "-G     1246   1246  37252a93-0b5e-4afc-a2f1-9b8c94677920    1\n",
       "-T        1      1  a371539f-4db9-47e5-8167-b92eee07a52a    1\n",
       "A-        4      4  b21d62dd-fb6a-49de-9269-8cd47e5812e5    1\n",
       "AA        2      2  0cb46d10-cda8-48cc-9692-1f362f3bf805    1\n",
       "AC        1      1  d1561e25-6b77-49ba-aefc-c35c8e37a4b2    1\n",
       "AG      258    258  3beba5ff-e3b2-4f39-8677-11c4df61286f    1\n",
       "C-        2      2  9881f62c-48c8-4a4a-9077-da998bfa285f    1\n",
       "CA        1      1  138f3737-5fda-4b56-8350-70dbc7043888    1\n",
       "CG      216    216  3f19c25c-4a69-489d-88ef-a0cb04afc557    1\n",
       "G-      911    911  7b649e13-b7e4-4000-9f78-dd86db465ddf    1\n",
       "GA      522    522  c2989e3e-0054-433f-8dab-41377f00cacf    1\n",
       "GC      102    102  22d216e6-0e56-42b8-9590-7c7c23a81e52    1\n",
       "GG    75812  75812  eaa26c97-60c1-4f84-a686-fe8ad76476d6    1\n",
       "GT       81     81  05223156-3110-4c9e-bcee-f663694bc863    1\n",
       "T-        1      1  49270610-2f46-4895-a22a-cc9bc51c38c7    1\n",
       "TA        1      1  08da06fe-6f62-4239-b6cd-62d2d1dece00    1\n",
       "TG       72     72  a2d695a2-5cb6-4810-8619-0c3fc4f652ed    1\n",
       "TT        1      1  c7cc8f74-5c39-4121-8314-7a1e08dca56e    1"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "myloc=[94,11965]\n",
    "bammy=pysam.AlignmentFile(\"Muscle/bam/190314_UCLA_young.rot.bam\", \"rb\")\n",
    "\n",
    "mapped=bammy.pileup(\"chrM_rot\", myloc[0], myloc[1],min_base_quality=0, max_depth=1000000)\n",
    "readvariants=dict()\n",
    "for pileupcolumn in mapped:\n",
    "    if pileupcolumn.pos in myloc:\n",
    "        print(\"coverage at base %s = %s\" %(pileupcolumn.pos, pileupcolumn.n))\n",
    "        for pileupread in pileupcolumn.pileups:\n",
    "            readname=pileupread.alignment.query_name\n",
    "            if pileupread.is_del:\n",
    "                base=\"-\" \n",
    "            else:\n",
    "                base=pileupread.alignment.query_sequence[pileupread.query_position]\n",
    "            if readname in readvariants:\n",
    "                readvariants[readname]=readvariants[readname]+base\n",
    "            else:\n",
    "                readvariants[readname]=base\n",
    "\n",
    "readvariants={key: value for (key, value) in readvariants.items() if len(value) == len(myloc)}\n",
    "\n",
    "\n",
    "geno=pd.DataFrame.from_dict(readvariants, orient='index',columns=[\"geno\"])\n",
    "geno.index.name = \"read\"\n",
    "geno.reset_index(inplace=True)\n",
    "geno.groupby(\"geno\").describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "coverage at base 94 = 13094\n",
      "coverage at base 11965 = 3693\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"4\" halign=\"left\">read</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>unique</th>\n",
       "      <th>top</th>\n",
       "      <th>freq</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>geno</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>--</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>7aaa30ff-54d9-465b-a504-6db04766f416</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-A</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>68cb0e16-8604-4569-8fa4-d5f0b50e6cf2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-G</th>\n",
       "      <td>81</td>\n",
       "      <td>81</td>\n",
       "      <td>69162c04-b289-4249-a006-6371ec6791de</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AG</th>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "      <td>0efe15de-f1e3-4d3b-8de3-bf02a3473ed0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CG</th>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>ecd6543c-cd69-4a73-ae6f-d167e86a2226</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G-</th>\n",
       "      <td>45</td>\n",
       "      <td>45</td>\n",
       "      <td>0dfa442f-21c4-43fc-a8b5-1808c0f25a91</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GA</th>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>e3262582-6218-4161-b07e-6f211cc13c21</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GG</th>\n",
       "      <td>2880</td>\n",
       "      <td>2880</td>\n",
       "      <td>4cfb7279-ccfe-4634-8a7e-8854a9ad45e3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GT</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>74bc276b-e9ef-4868-a562-2ec3080157c4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TG</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>783b676f-c4f0-4e6f-9e1f-68f6ccae2cfa</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      read                                                  \n",
       "     count unique                                   top freq\n",
       "geno                                                        \n",
       "--       2      2  7aaa30ff-54d9-465b-a504-6db04766f416    1\n",
       "-A       1      1  68cb0e16-8604-4569-8fa4-d5f0b50e6cf2    1\n",
       "-G      81     81  69162c04-b289-4249-a006-6371ec6791de    1\n",
       "AG      20     20  0efe15de-f1e3-4d3b-8de3-bf02a3473ed0    1\n",
       "CG      12     12  ecd6543c-cd69-4a73-ae6f-d167e86a2226    1\n",
       "G-      45     45  0dfa442f-21c4-43fc-a8b5-1808c0f25a91    1\n",
       "GA      23     23  e3262582-6218-4161-b07e-6f211cc13c21    1\n",
       "GG    2880   2880  4cfb7279-ccfe-4634-8a7e-8854a9ad45e3    1\n",
       "GT       1      1  74bc276b-e9ef-4868-a562-2ec3080157c4    1\n",
       "TG       1      1  783b676f-c4f0-4e6f-9e1f-68f6ccae2cfa    1"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "myloc=[94,11965]\n",
    "bammy=pysam.AlignmentFile(\"Muscle/bam/210506_y3_UCLA.rot.bam\", \"rb\")\n",
    "\n",
    "mapped=bammy.pileup(\"chrM_rot\", myloc[0], myloc[1],min_base_quality=0, max_depth=1000000)\n",
    "readvariants=dict()\n",
    "for pileupcolumn in mapped:\n",
    "    if pileupcolumn.pos in myloc:\n",
    "        print(\"coverage at base %s = %s\" %(pileupcolumn.pos, pileupcolumn.n))\n",
    "        for pileupread in pileupcolumn.pileups:\n",
    "            readname=pileupread.alignment.query_name\n",
    "            if pileupread.is_del:\n",
    "                base=\"-\" \n",
    "            else:\n",
    "                base=pileupread.alignment.query_sequence[pileupread.query_position]\n",
    "            if readname in readvariants:\n",
    "                readvariants[readname]=readvariants[readname]+base\n",
    "            else:\n",
    "                readvariants[readname]=base\n",
    "\n",
    "readvariants={key: value for (key, value) in readvariants.items() if len(value) == len(myloc)}\n",
    "\n",
    "\n",
    "geno=pd.DataFrame.from_dict(readvariants, orient='index',columns=[\"geno\"])\n",
    "geno.index.name = \"read\"\n",
    "geno.reset_index(inplace=True)\n",
    "geno.groupby(\"geno\").describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "coverage at base 94 = 90930\n",
      "coverage at base 11965 = 37474\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"4\" halign=\"left\">read</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>unique</th>\n",
       "      <th>top</th>\n",
       "      <th>freq</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>geno</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>--</th>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>05dce4dc-4833-49ca-a467-052db1d87524</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-A</th>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>aeb11513-1cee-42b7-9507-4bc969a17d6e</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-G</th>\n",
       "      <td>398</td>\n",
       "      <td>398</td>\n",
       "      <td>2141c9ad-138d-4f4e-9e23-9d65bd05750f</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-T</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>d85362ae-4a36-4b5f-9885-74d3eefaee70</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AA</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>21ed31a2-c4e6-47bb-bcf1-576719c11a04</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AG</th>\n",
       "      <td>110</td>\n",
       "      <td>110</td>\n",
       "      <td>3750778a-c7a6-4fbb-8325-2201d4f9dfc9</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CG</th>\n",
       "      <td>113</td>\n",
       "      <td>113</td>\n",
       "      <td>a58011cd-13f8-4ca0-8fae-bda322bfa9d5</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G-</th>\n",
       "      <td>368</td>\n",
       "      <td>368</td>\n",
       "      <td>88cd95b5-2dfb-4ea4-92d8-ac57f5062737</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GA</th>\n",
       "      <td>237</td>\n",
       "      <td>237</td>\n",
       "      <td>da6137e8-4bef-4391-816f-fbe5ba4ce2e9</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GC</th>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>87582883-0e49-4b91-98a2-413eab2afdf4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GG</th>\n",
       "      <td>21451</td>\n",
       "      <td>21451</td>\n",
       "      <td>482be819-6a3f-4d09-9adf-ce4158c1fa93</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GT</th>\n",
       "      <td>28</td>\n",
       "      <td>28</td>\n",
       "      <td>87bdf581-87a7-4ac5-aa37-d96e4fadfdcb</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TG</th>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>36a899d0-932b-46f0-bc19-bdc7cb3536a2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       read                                                  \n",
       "      count unique                                   top freq\n",
       "geno                                                         \n",
       "--        8      8  05dce4dc-4833-49ca-a467-052db1d87524    1\n",
       "-A        9      9  aeb11513-1cee-42b7-9507-4bc969a17d6e    1\n",
       "-G      398    398  2141c9ad-138d-4f4e-9e23-9d65bd05750f    1\n",
       "-T        1      1  d85362ae-4a36-4b5f-9885-74d3eefaee70    1\n",
       "AA        1      1  21ed31a2-c4e6-47bb-bcf1-576719c11a04    1\n",
       "AG      110    110  3750778a-c7a6-4fbb-8325-2201d4f9dfc9    1\n",
       "CG      113    113  a58011cd-13f8-4ca0-8fae-bda322bfa9d5    1\n",
       "G-      368    368  88cd95b5-2dfb-4ea4-92d8-ac57f5062737    1\n",
       "GA      237    237  da6137e8-4bef-4391-816f-fbe5ba4ce2e9    1\n",
       "GC       13     13  87582883-0e49-4b91-98a2-413eab2afdf4    1\n",
       "GG    21451  21451  482be819-6a3f-4d09-9adf-ce4158c1fa93    1\n",
       "GT       28     28  87bdf581-87a7-4ac5-aa37-d96e4fadfdcb    1\n",
       "TG       12     12  36a899d0-932b-46f0-bc19-bdc7cb3536a2    1"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "myloc=[94,11965]\n",
    "bammy=pysam.AlignmentFile(\"Muscle/bam/old3_UCLA_merge.rot.bam\", \"rb\")\n",
    "\n",
    "mapped=bammy.pileup(\"chrM_rot\", myloc[0], myloc[1],min_base_quality=0, max_depth=1000000)\n",
    "readvariants=dict()\n",
    "for pileupcolumn in mapped:\n",
    "    if pileupcolumn.pos in myloc:\n",
    "        print(\"coverage at base %s = %s\" %(pileupcolumn.pos, pileupcolumn.n))\n",
    "        for pileupread in pileupcolumn.pileups:\n",
    "            readname=pileupread.alignment.query_name\n",
    "            if pileupread.is_del:\n",
    "                base=\"-\" \n",
    "            else:\n",
    "                base=pileupread.alignment.query_sequence[pileupread.query_position]\n",
    "            if readname in readvariants:\n",
    "                readvariants[readname]=readvariants[readname]+base\n",
    "            else:\n",
    "                readvariants[readname]=base\n",
    "\n",
    "readvariants={key: value for (key, value) in readvariants.items() if len(value) == len(myloc)}\n",
    "\n",
    "\n",
    "geno=pd.DataFrame.from_dict(readvariants, orient='index',columns=[\"geno\"])\n",
    "geno.index.name = \"read\"\n",
    "geno.reset_index(inplace=True)\n",
    "geno.groupby(\"geno\").describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "coverage at base 94 = 25933\n",
      "coverage at base 11965 = 13890\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"4\" halign=\"left\">read</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>unique</th>\n",
       "      <th>top</th>\n",
       "      <th>freq</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>geno</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>--</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>b09454c5-b0e9-455f-a625-f097ef9d9a9d</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-A</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>d080722b-d75e-44bf-af41-a74013a446c8</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>-G</th>\n",
       "      <td>251</td>\n",
       "      <td>251</td>\n",
       "      <td>1e92f113-3b3f-4eea-a3e9-621b71df8355</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A-</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>be652b4d-4f52-4e9e-8cdc-204f18ce4ef2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AA</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>86654cd1-9254-49d9-8b53-63d26866eac9</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AG</th>\n",
       "      <td>69</td>\n",
       "      <td>69</td>\n",
       "      <td>fa01ee8a-aebb-408d-95f0-7292aaed6dba</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C-</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>43ecf77a-0099-4123-a328-16d4caeef945</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CA</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4a4c6c46-4e0b-4ce4-824d-f341a9a1d414</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CG</th>\n",
       "      <td>52</td>\n",
       "      <td>52</td>\n",
       "      <td>c79dd195-4e7e-4745-b28c-9e7c675ed7ac</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G-</th>\n",
       "      <td>205</td>\n",
       "      <td>205</td>\n",
       "      <td>2a462d40-ea23-4af2-b525-af94f96ee7d7</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GA</th>\n",
       "      <td>112</td>\n",
       "      <td>112</td>\n",
       "      <td>474d9c73-3af7-4ba6-ad17-09041cee123c</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GC</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0b9111ba-0b9e-4387-a3fa-d2cb70e37674</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GG</th>\n",
       "      <td>11001</td>\n",
       "      <td>11001</td>\n",
       "      <td>85e6d021-8163-435f-8290-76bd07f64ef0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GT</th>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>43d7a55f-90a7-40a5-b785-d9fa4f836bf1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TG</th>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>d0dc6b6b-9697-493c-afed-af926ceed740</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       read                                                  \n",
       "      count unique                                   top freq\n",
       "geno                                                         \n",
       "--        5      5  b09454c5-b0e9-455f-a625-f097ef9d9a9d    1\n",
       "-A        3      3  d080722b-d75e-44bf-af41-a74013a446c8    1\n",
       "-G      251    251  1e92f113-3b3f-4eea-a3e9-621b71df8355    1\n",
       "A-        3      3  be652b4d-4f52-4e9e-8cdc-204f18ce4ef2    1\n",
       "AA        1      1  86654cd1-9254-49d9-8b53-63d26866eac9    1\n",
       "AG       69     69  fa01ee8a-aebb-408d-95f0-7292aaed6dba    1\n",
       "C-        2      2  43ecf77a-0099-4123-a328-16d4caeef945    1\n",
       "CA        1      1  4a4c6c46-4e0b-4ce4-824d-f341a9a1d414    1\n",
       "CG       52     52  c79dd195-4e7e-4745-b28c-9e7c675ed7ac    1\n",
       "G-      205    205  2a462d40-ea23-4af2-b525-af94f96ee7d7    1\n",
       "GA      112    112  474d9c73-3af7-4ba6-ad17-09041cee123c    1\n",
       "GC        3      3  0b9111ba-0b9e-4387-a3fa-d2cb70e37674    1\n",
       "GG    11001  11001  85e6d021-8163-435f-8290-76bd07f64ef0    1\n",
       "GT       14     14  43d7a55f-90a7-40a5-b785-d9fa4f836bf1    1\n",
       "TG        8      8  d0dc6b6b-9697-493c-afed-af926ceed740    1"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "myloc=[94,11965]\n",
    "bammy=pysam.AlignmentFile(\"Muscle/bam/old4_UCLA_merge.rot.bam\", \"rb\")\n",
    "\n",
    "mapped=bammy.pileup(\"chrM_rot\", myloc[0], myloc[1],min_base_quality=0, max_depth=1000000)\n",
    "readvariants=dict()\n",
    "for pileupcolumn in mapped:\n",
    "    if pileupcolumn.pos in myloc:\n",
    "        print(\"coverage at base %s = %s\" %(pileupcolumn.pos, pileupcolumn.n))\n",
    "        for pileupread in pileupcolumn.pileups:\n",
    "            readname=pileupread.alignment.query_name\n",
    "            if pileupread.is_del:\n",
    "                base=\"-\" \n",
    "            else:\n",
    "                base=pileupread.alignment.query_sequence[pileupread.query_position]\n",
    "            if readname in readvariants:\n",
    "                readvariants[readname]=readvariants[readname]+base\n",
    "            else:\n",
    "                readvariants[readname]=base\n",
    "\n",
    "readvariants={key: value for (key, value) in readvariants.items() if len(value) == len(myloc)}\n",
    "\n",
    "\n",
    "geno=pd.DataFrame.from_dict(readvariants, orient='index',columns=[\"geno\"])\n",
    "geno.index.name = \"read\"\n",
    "geno.reset_index(inplace=True)\n",
    "geno.groupby(\"geno\").describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Now load in illumina data, look at % of each snp\n",
    "cd Timp_data/HV/illumina"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref=\"/Users/amyvandiver/Box/Nanopore/TimpLab/Ref/GRCh38.primary_assembly.genome.fa\"\n",
    "fq1=\"mito_S1_L001_R1_001.fastq\"\n",
    "fq2=\"mito_S1_L001_R2_001.fastq\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!echo $(( $(wc -l < $fq1) / 4 )) reads\n",
    "!echo $(( $(wc -l < $fq2) / 4 )) reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bwa index {ref}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bwa mem {ref} {fq1} {fq2}| samtools view -bS -uF 4 | samtools sort -o mito_S1.bam\n",
    "!samtools index mito_S1.bam\n",
    "!samtools idxstats mito_S1.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bam= \"mito_S1.bam\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pysam\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "snp1=[1641]\n",
    "snp2=[13512]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "samfile = pysam.AlignmentFile(bam, \"rb\" )\n",
    "variants1=dict()\n",
    "for pileupcolumn in samfile.pileup(\"chrM\",max_depth=1000000):\n",
    "    if pileupcolumn.pos in snp1:\n",
    "        print (\"\\ncoverage at base %s = %s\" %\n",
    "           (pileupcolumn.pos, pileupcolumn.n))\n",
    "        for pileupread in pileupcolumn.pileups:\n",
    "            readname=pileupread.alignment.query_name\n",
    "            if not pileupread.is_del and not pileupread.is_refskip:\n",
    "                # query position is None if is_del or is_refskip is set.\n",
    "                base=pileupread.alignment.query_sequence[pileupread.query_position]\n",
    "                variants1[readname]=base\n",
    "samfile.close()\n",
    "\n",
    "geno=pd.DataFrame.from_dict(variants1, orient='index',columns=[\"geno\"])\n",
    "geno.index.name = \"read\"\n",
    "geno.reset_index(inplace=True)\n",
    "geno.groupby(\"geno\").describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "geno.to_csv(\"geno_1642.csv\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "samfile = pysam.AlignmentFile(bam, \"rb\" )\n",
    "variants1=dict()\n",
    "for pileupcolumn in samfile.pileup(\"chrM\",max_depth=25000):\n",
    "    if pileupcolumn.pos in snp2:\n",
    "        print (\"\\ncoverage at base %s = %s\" %\n",
    "           (pileupcolumn.pos, pileupcolumn.n))\n",
    "        for pileupread in pileupcolumn.pileups:\n",
    "            readname=pileupread.alignment.query_name\n",
    "            if not pileupread.is_del and not pileupread.is_refskip:\n",
    "                # query position is None if is_del or is_refskip is set.\n",
    "                base=pileupread.alignment.query_sequence[pileupread.query_position]\n",
    "                variants1[readname]=base\n",
    "samfile.close()\n",
    "\n",
    "geno=pd.DataFrame.from_dict(variants1, orient='index',columns=[\"geno\"])\n",
    "geno.index.name = \"read\"\n",
    "geno.reset_index(inplace=True)\n",
    "geno.groupby(\"geno\").describe()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "geno.to_csv(\"geno_13513.csv\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
